Source/TCPConnection.swift

```diff
@@ -30,145 +30,97 @@ public final class TCPConnection: Connection {
+        try ensureStreamIsOpen()
 
         var data = Data.buffer(with: byteCount)
-        let bytesProcessed = data.withUnsafeMutableBufferPointer {
+        let received = data.withUnsafeMutableBufferPointer {
             tcprecvlh(socket, $0.baseAddress, 1, $0.count, deadline.int64milliseconds)
         }
 
-        try TCPError.assertNoReceiveErrorWithData(data, bytesProcessed: bytesProcessed)
-        return Data(data.prefix(bytesProcessed))
-    }
+        let receivedData = Data(data.prefix(received))
 
-    public func receive(from start: Int, to end: Int, timingOut deadline: Double = .never) throws -> Data {
-        let socket = try getSocket()
-        try assertNotClosed()
-
-        if start <= 0 || end <= 0 {
-            throw TCPError.unknown(description: "Marks should be > 0")
-        }
-
-        if start > end {
-            throw TCPError.unknown(description: "loweWaterMark should be less than highWaterMark")
+        do {
+            try ensureLastOperationSucceeded()
+        } catch {
+            throw TCPError.failedToReceiveCompletely(received: receivedData)
         }
 
-        var data = Data.buffer(with: end)
-        let bytesProcessed = data.withUnsafeMutableBufferPointer {
-            tcprecvlh(socket, $0.baseAddress, start, $0.count, deadline.int64milliseconds)
-        }
-
-        try TCPError.assertNoReceiveErrorWithData(data, bytesProcessed: bytesProcessed)
-        return Data(data.prefix(bytesProcessed))
+        return receivedData
     }
 
-    public func receive(upTo byteCount: Int, until delimiter: String, timingOut deadline: Double = .never) throws -> Data {
+    public func close() throws {
         let socket = try getSocket()
-        try assertNotClosed()
-
-
-        var data = Data.buffer(with: byteCount)
-        let bytesProcessed = data.withUnsafeMutableBufferPointer {
-            tcprecvuntil(socket, $0.baseAddress, $0.count, delimiter, delimiter.utf8.count, deadline.int64milliseconds)
-        }
-
-        try TCPError.assertNoReceiveErrorWithData(data, bytesProcessed: bytesProcessed)
-        return Data(data.prefix(bytesProcessed))
-    }
-
-    public func close() -> Bool {
-        guard let socket = self.socket else {
-            closed = true
-            return true
-        }
 
         if closed {
-            return false
+            throw ClosableError.alreadyClosed
         }
 
-        closed = true
         tcpclose(socket)
-        return true
+        try ensureLastOperationSucceeded()
+        closed = true
     }
 
     private func getSocket() throws -> tcpsock {
         guard let socket = self.socket else {
-            throw TCPError.closedSocket(description: "Connection has not been initialized. You must first open to the connection.")
-        }
-        if socket == nil {
-            throw TCPError.closedSocket(description: "Connection has not been initialized. You must first open to the connection.")
+            throw SystemError.socketIsNotConnected
```

Comment:
If we remove the nil check implementations will crash with a EXEC_BAD_ACCESS error.